runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DockerJobBundleFactory.java

```diff
@@ -69,6 +70,21 @@
 public class DockerJobBundleFactory implements JobBundleFactory {
   private static final Logger LOG = LoggerFactory.getLogger(DockerJobBundleFactory.class);
 
+  /** Factory that creates {@link JobBundleFactory} for the given {@link JobInfo}. */
+  public interface JobBundleFactoryFactory {
+    DockerJobBundleFactory create(JobInfo jobInfo) throws Exception;
+  }
+  // TODO: a hacky way to override the factory for testing.
+  // Should be replaced with mechanism that let's users configure their own factory
```

Comment:
As a future thought: We should expect that the environment proto that we are passing around is changed so that it has different types, one which is `docker`, another one which is `inprocess` and then we can use a selector to choose which environment factory to use within Flink/Spark/...

Reply:
Right, that would require changes in this area to pass on the environment: https://github.com/apache/beam/blob/1207c28ee3cb38b9e353fc537dea7abe296480ed/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/BatchFlinkExecutableStageContext.java#L90